storagenode/piecestore/endpoint.go

```diff
@@ -335,7 +338,13 @@ func (endpoint *Endpoint) Download(stream pb.Piecestore_DownloadServer) (err err
 
 	pieceReader, err := endpoint.store.Reader(ctx, limit.SatelliteId, limit.PieceId)
 	if err != nil {
-		return ErrInternal.Wrap(err) // TODO: report grpc status internal server error
+		if os.IsNotExist(err) {
+			return status.Error(codes.NotFound, err.Error())
+		}
+		if os.IsPermission(err) {
+			return status.Error(codes.PermissionDenied, err.Error())
```

Comment:
I'm not convinced that if the permission of the files doesn't allow to read the the piece the storagenode has to return this GRPC error,  because we don't use the file permission to bock the access to pieces and if that happens is because of a bad operation of the storage node operator.

For a hones node, which should run this code, for me it should be also `NotFound`.

Reply:
Yes, you are right. Now I am re-reading the godoc for `PermissionDenied` and it says:
> PermissionDenied indicates the caller does not have permission to execute the specified operation.

It's really not the caller (the satellite or the uplink) that does not have permission to access the piece. It's a problem on the storage node that the piece cannot be opened.

However, I also don't think that `NotFound` is correct too. Because the piece is there, but it cannot be opened because of an issue that can be fixed. How about returning `FailedPrecondition` or `Internal`? Which one would you choose?

Reply:
I agree, `NotFound` isn't good in comparison of any of the 2 that you used, however I didn't check the list of codes yesterday and I only based my reply on those 2 error codes that you were using.

Between those 2, I think that `Internal` is more appropriated because while the docs don't say that the client can do whatsoever other operation to change the state of the system for fulfilling the precondition, I think that failed preconditions are based on correct system states, so due to other operations (by the same client, other clients or even by the usual flow of the system) the system will meet the precondition at some point. For this specific case, the permissions should never have changed by our system (I'm not considering that the cause of the change is due to a bug), it has been an external factor which has changed them so I believe that the description of `Internal` fits to it

> Internal errors. This means that some invariants expected by the underlying system have been broken. This error code is reserved for serious errors.

Reply:
Deal! It will be `Internal` :)